src/QsCompiler/Transformations/CallGraphWalker.cs

```diff
@@ -0,0 +1,344 @@
+        public List<ImmutableArray<CallGraphNode>> GetCallCycles()
+        {
+            var callStack = new Dictionary<CallGraphNode, CallGraphNode>();
+            var finished = new HashSet<CallGraphNode>();
+            var cycles = new List<ImmutableArray<CallGraphNode>>();
+
+            void processDependencies(CallGraphNode node)
+            {
+                if (_Dependencies.TryGetValue(node, out var dependencies))
+                {
+                    foreach (var (curr, _) in dependencies)
+                    {
+                        if (!finished.Contains(curr))
+                        {
+                            if (callStack.TryGetValue(curr, out var next))
+                            {
+                                // Cycle detected
+                                var cycle = new List<CallGraphNode>() { curr };
+                                if (!curr.Equals(next)) // If the cycle is a direct recursion, we only want the one node
+                                {
+                                    do
+                                    {
+                                        cycle.Add(next);
+                                    } while (callStack.TryGetValue(next, out next));
+                                }
+
+                                cycles.Add(cycle.ToImmutableArray());
+                            }
+                            else
+                            {
+                                callStack[node] = curr;
+                                processDependencies(curr);
+                                callStack.Remove(node);
+                            }
+                        }
+                    }
+                }
+
+                finished.Add(node);
+            }
+
+            foreach (var key in _Dependencies.Keys)
+            {
+                if (!finished.Contains(key))
+                {
+                    processDependencies(key);
+                }
+            }
+
+            return cycles;
+        }
+    }
+
+    /// <summary>
+    /// This transformation walks through the compilation without changing it, building up a call graph as it does.
+    /// This call graph is then returned to the user.
+    /// </summary>
+    public static class BuildCallGraph
+    {
+        public static CallGraph Apply(QsCompilation compilation)
+        {
+            var walker = new BuildGraph();
+
+            foreach (var ns in compilation.Namespaces)
+            {
+                walker.Namespaces.OnNamespace(ns);
+            }
+
+            return walker.SharedState.graph;
+        }
+
+        private class BuildGraph : SyntaxTreeTransformation<BuildGraph.TransformationState>
+        {
+            public class TransformationState
+            {
+                internal QsSpecialization spec;
+
+                internal bool inCall = false;
+                internal bool adjoint = false;
+                internal bool controlled = false;
+
```

Comment:
Any good reason not to use QsSpecializationKind here?

Reply:
With the adjoint, it is easier to flip back and forth a Boolean than to change back and forth a QsSpecializationKind.QsBody to QsSpecializationKind.QsAdjoint or QsSpecializationKind.QsControlled to QsSpecializationKind.QsControlledAdjoint as we encounter adjoint functor applications. It is easier to have all that logic done at the end, when we are not going to encounter any more functor applications. I will rename the fields to make it clearer that these are in reference to functor applications and not what specialization of a callable we are in.